app/lib/test_executor.rb

```diff
@@ -0,0 +1,66 @@
+class TestExecutor
+  HOST_WORKDIR = '/usr/src/app'
+  WORKDIR = '/tmp/sandbox'
+
+  def self.images
+    @images ||= {}
+  end
+
+  def self.run_tests(submission)
+    problem = submission.problem
```

Comment:
You want to eager load all the tests for a problem. Otherwise it makes `N+1` queries in the next loop.